### PR TITLE
Refactor pattern matching [cps/tasks/download.py: parse yt-dlp output for download progress percentage]

### PR DESCRIPTION
**Pull request by deldesir**
These changes enhance the download task in several ways:

   - The download progress is now calculated dynamically based on the subprocess output
   - Updated the pattern to identify progress updates more reliably during the download process.
   - Cleaned up the code by removing redundant lines. Also, added the missing import statement for the 're' module.
   
---

Smoke and feature tested on Ubuntu 24.04

- [x] I reviewed the changes in the download task logic and subprocess handling.

- [x] I confirm that the dynamic progress calculation reflects the download progress.

- [x] I verified the improved subprocess output pattern for reliable progress tracking.

- [x] I checked the adjusted completion condition for 99% progress at 100% subprocess completion

Testing Steps in Calibre-Web UI
- Initiate a download task for a YouTube url.
- Monitor the progress updates in the tasks' view.
- Verify that the progress is accurately calculated and reaches 100% upon successful completion.

### CODE REVIEW

File: cps/tasks/download.py
Comment (holta) [inline cps/tasks/download.py:1]: @deldesir Special Request:

Please communicate in 3-6 words the central purpose of each commit.

Not in the hidden secondary commit message.  But in the actual commit message.

When non-descriptive commit messages like "Update download.py" it's becomes extremely hard for everyone to understand (1) what is happening and (2) why. 😕

Goal: Commits Messages should Communicate Clear Intent! ✅

Comment (deldesir) [reply cps/tasks/download.py:1]: My bad, I thought I wrote "Fix indentation"

Comment (deldesir) [reply cps/tasks/download.py:1]: I see, I wrote it in description field. I thought I was putting the commit message in the title. It was an accident.

Comment (holta) [reply cps/tasks/download.py:1]: Everyone benefits when motivation is clear, concisely e.g. anything like:

- "cps/tasks/download.py: Fix indentation"

- "downloaded.py: Fix indenting in ~30 lines"

(Or anything similar!)

**Comment by deldesir** [comment]
>     * [x]  I checked the adjusted completion condition for 99% progress at 100% subprocess completion

FYI, I had to record the first 100% progress to %99 so the script won't stop before yt-dlp finishes. There is often one or two more quick 0%-100% cycles after the first one (thubmnail and/or subtitle downloads)

File: cps/tasks/download.py [hunk 1]
```diff
@@ -41,44 +42,42 @@ def run(self, worker_thread):
...
-                            response = requests.get(self.original_url, params={"requested_files": requested_files, "current_user_name": self.current_user_name, "shelf_title": shelf_title})
-                            if response.status_code == 200:
-                                log.info("Successfully sent the list of requested files to %s", self.original_url)
+                        if pattern_progress in line:
+                            percentage = int(re.search(r'\d+', line).group())
+                            if percentage < 100:
+                                self.progress = percentage / 100
                             else:
-                                log.error("Failed to send the list of requested files to %s", self.original_url)
+                                self.progress = 0.99
```
Comment (deldesir) [inline cps/tasks/download.py:55]: FYI, I record the first 100% progress cycle as %99 so the script won't stop before yt-dlp finishes. There is often one or more quick 0%-100% cycles after the first one (for example the thumbnail and/or subtitles downloads)

**Comment by holta** [comment]
Has this PR been quickly tested?

If so we should feel free to merge it, to keep community momentum!

File: cps/tasks/download.py (hunk 1 above)
Comment (holta) [reply cps/tasks/download.py:55]: I definitely don't understand it yet, ah well 😎 

But thanks so much for explaining!

Comment (holta) [reply cps/tasks/download.py:55]: > FYI, I record the first 100% progress cycle as %99 so the script won't stop before yt-dlp finishes. There is often one or more quick 0%-100% cycles after the first one (for example the thumbnail and/or subtitles downloads)

@deldesir ideally put some-or-all of this explanation in-line as a comment, near the Python code in question?

**Comment by deldesir** [comment]
Yes, it is well tested.
> Smoke and feature tested on Ubuntu 24.04
> 
> * [x]  I reviewed the changes in the download task logic and subprocess handling.
> * [x]  I confirm that the dynamic progress calculation reflects the download progress.
> * [x]  I verified the improved subprocess output pattern for reliable progress tracking.
> * [x]  I checked the adjusted completion condition for 99% progress at 100% subprocess completion
> 
> Testing Steps in Calibre-Web UI
> 
> * Initiate a download task for a YouTube url.
> * Monitor the progress updates in the tasks' view.
> * Verify that the progress is accurately calculated and reaches 100% upon successful completion.

File: cps/tasks/download.py (hunk 1 above)
Comment (holta) [reply cps/tasks/download.py:55]: Please put this explanation (or similar) into a comment near the actual Python code.

Then let's move forward with merging this PR.

Thanks!

**Comment by deldesir** [comment]
https://github.com/iiab/calibre-web/blob/4da5c46032f251cf3cca000608a55f8edd320a42/cps/tasks/download.py#L47-L58
The first 100% progress cycle is recorded as %99. This prevent the script from being stopped before yt-dlp finishes. There is often one or more quick 0%-100% cycles after the first one (for example the thumbnail and/or subtitles downloads)

File: cps/tasks/download.py [hunk 2]
```diff
@@ -41,44 +42,42 @@ def run(self, worker_thread):
...
-                            shelf_title = None
-                        conn.close()
-
-                        if self.original_url:
-                            response = requests.get(self.original_url, params={"requested_files": requested_files, "current_user_name": self.current_user_name, "shelf_title": shelf_title})
-                            if response.status_code == 200:
-                                log.info("Successfully sent the list of requested files to %s", self.original_url)
+                        if pattern_progress in line:
+                            percentage = int(re.search(r'\d+', line).group())
+                            if percentage < 100:
```
Comment (holta) [inline cps/tasks/download.py:52]: Let's include more in-line comments contextualizing what's happening and why.

Something like the following for starters?

```suggestion
                            # 2024-01-10: 99% (a bit arbitrary) is explained here...
                            # https://github.com/iiab/calibre-web/pull/88#issuecomment-1885916421
                            if percentage < 100:
```

File: cps/tasks/download.py [hunk 3]
```diff
@@ -41,44 +42,42 @@ def run(self, worker_thread):
                 p = process_open(subprocess_args, newlines=True)
 
                 # Define the pattern for the subprocess output
-                pattern_analyze = r"Running ANALYZE"
-                pattern_download = r"'action': 'download'"
+                pattern_progress = r"downloading"
```
[the remark this answers is not in the document]
Comment (deldesir) [reply cps/tasks/download.py:45]: I agree

**Comment by holta** [comment]
Legibility = Understandability needs to be the central driving force going forward:

- Sweeping indentation changes make PR's almost impossible to read :x: 

- Large-scale indentation changes need to be moved to a completely separate PE in future :heavy_check_mark: 

_Thank @deldesir for taking approachable (small) PR's very seriously!_

_AI² = Articulate Assumptions, Intent & Intuition_

**Comment by holta** [comment]
@deldesir any quick tip(s) you can provide to @EMG70 to help him reconfirm / validate this just-merged PR?

e.g. Is there anything specific he should look for in "Tasks" view today or in coming days?

https://github.com/iiab/calibre-web/blob/d594aeba06ab91e684122ca03062564d71f012e5/cps/tasks/download.py#L44-L46

**Comment by deldesir** [comment]
> @deldesir any quick tip(s) you can provide to @EMG70 to help him reconfirm / validate this just-merged PR?

A positive scenario testing should finish with 100% recorded in tasks' progress column.

> e.g. Is there anything specific he should look for in "Tasks" view today or in coming days?

We will work today on tasks'  messages printed in Task column (this is a discussion for another PR)

File: cps/tasks/download.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (holta) [reply cps/tasks/download.py]: Could use additional testing!  Thanks @deldesir and @EMG70 if you can reconfirm Line 46 (now merged) is working well!

[the remark this answers is not in the document]
Comment (holta) [reply cps/tasks/download.py]: Bad News:

Regex `"^downloading"` was being used incorrectly yesterday.  Functional testing didn't happen when it should have (to catch this serious oversight). :x:

Good News:

PR's like #89 and #93 are getting much **smaller and hence tangibly clearer** ~ "TDD Craftsmanship" Organic + Incremental + Visible ~ nurturing a Responsive Culture of Collaboration.

_By learning from our mistakes (the only way!) one day at a time_ :seedling: :seedling: :seedling:

**Pull request merged**